### DISCUSSION TITLE
Allow Users to Select Displayed Columns in Grid View

### DISCUSSION THREAD

**Discussion by kukuk52** [cortexrd/Knack-Toolkit-Library#45] [Ideas]
It would be great if the KTL would allow users to select the columns that they'd like to display in a given grid (much like you've done with the grid filters). Saving your preferred column selection locally would allow quick access to displays of preferred data depending on the project the user is working on. Ideally, this would also apply when exporting that grid (although I realize that's a much heavier lift).

**Comment by cortexrd** [comment]
Hi Daniel,

Thanks for suggesting this feature.  I've brainstormed on it lately and I think I have a good idea how I'd do this.

To prevent conflicting with the User/Public Filters' real-estate, I think I'd use a drop-down or a pop-up window to select the "Layout" instead of horizontal buttons like filters. Since column toggling is not a feature you'd use 20 times a day most likely, that would be acceptable.

So you'd have new a "Hide/Show Columns" button that, when clicked, would...
1. open a dropdown or pop-up, where you'd see Save, Delete, Rename, Layout1, Layout2, etc.
2. add an eye icon (button) on each column, that would be green or red (for show/hide). When the eye icon is clicked, it would toggle the column's visibility.
3. The layouts would be saved, per application, user and view.  They would be saved in localStorage for persistence.

But the export to CSV only for the visible columns... that's another story indeed. That would require re-writing the export, with page flipping when data is more than 1000 records.

Any other ideas?

Norm

**Comment by kukuk52** [comment]
Norm, this all sounds really great. I agree, a dropdown or pop-up would keep the page clean and allow the users to select and save their preferred views. I know this may be a bit more complicated, but have you considered universal preset column configurations (a developer or user admin could set up some specific views that users would have available on their end)?  

While the export function would really bring this together, I don't think it's something that should prevent or slow the development of this tool.  

Thanks again.

**Comment by kukuk52** [comment]
Hey Norm, 
Have you further considered implementing this feature? If so, when would you think it might be available? Thanks!

**Comment by cortexrd** [comment]
Hi Dan,

Yes that feature (call it a project) is still very much in the scope, and quite desirable.  Many people have asked for it actually, along with the Excel/CSV export based on visible columns only.

I have no time at the moment, being super busy with my main client's projects.  But I may have a contributor that could help me on this - let me check.

Norm

**Comment by cortexrd** [comment]
Hi Dan,

I've completed a part of the feature.  You can hide/show columns but not save the settings for now.

You can try it out to see a preview.  It's not officially released yet, so you need to set the version to **dev** in the loader, like this:

```
KnackInitAsync = function ($, callback) { // Load the Knack Toolkit Library (KTL)
    (window.LazyLoad = LazyLoad) && LazyLoad.js(['https://ctrnd.s3.amazonaws.com/Lib/KTL/KTL_Start.js'], () => {
        loadKtl($, callback, (typeof KnackApp === 'function' ? KnackApp : null), 'dev' /*KTL version, leave blank to get latest*/, 'full'/*min or full*/);
    })
};
```

Then do a Ctrl+F5 to force an update, bypassing the cache.

You will see a pale left-pointing caret (arrow) in each column header.  When you click on it, the columns is collapsed to 20 pixels and show with a hatching.  Click anywhere in the column to expand it back to normal.

Let me know if you like it,
Norm

**Comment by cortexrd** [comment]
Hi Dan,

Any feedback on the **_hsc** keyword?
It's now in official release **v0.18.8**.

Norm

**Reply by kukuk52** [reply]
Hi Norm, 

This is a great start and a useful tool that I'll be adding to a few of my grid views. That being said, I'd still really like to see something more like you described above where there would be a pop-up that would allow the user to select what columns they'd like to show/hide. To really make this useful, I think the selection needs to be saved with the logged-in user or on the local storage so that someone doesn't have to re-filter their view every time they navigate away from the page. A default selection would also be great if possible. 

I've been thinking through some workarounds to accomplish this, but to have it built right into the KTL would be amazing. 

Thanks again for all you're doing!

**Comment by kukuk52** [comment]
Would it be possible to use the _hc keyword and some conditions to make this work? If the user updated a record with a multi-select multiple choice field that contained the names of all the fields that were available to show/hide in a given grid could that be used as a reference to show/hide the various columns?

**Reply by CSWinnall** [reply]
I think that's a good idea and might be possible if you had a separate _hc for each column you want to hide and use ktlCond with the contains operator. Yes I think this could work.  The field with the list of columns could be in details view with an edit form the user could add or remove columns.

I will have to give this a go and get back to you.

Great idea.

Craig

**Reply by CSWinnall** [reply]
You could have a grid, in each row a name of a column and have a yes/no field and use the quick toggle keyword so the user could quickly toggle the columns on or off.

**Reply by cortexrd** [reply]
Indeed, very clever!  You're using the KTL in ways I didn't even think of myself LOL!
That should work for sure, with  Details view that has the checkboxes of columns to show, and several **_hc** in the grid, with **ktlCond**.

Norm

**Comment by kukuk52** [comment]
That's a pretty neat idea. What would the keyword input look like on the
main grid (the one where we'd want to show/hide columns) to get this to
work?

On Tue, Dec 5, 2023 at 3:38 PM CSWinnall ***@***.***> wrote:

> You could have a grid, in each row a name of a column and have a yes/no
> field and use the quick toggle keyword so the user could quickly toggle the
> columns on or off.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cortexrd/Knack-Toolkit-Library/discussions/45#discussioncomment-7768806>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANQJOGOONPHUFLHSPCYUUN3YH6A6FAVCNFSM6AAAAAAY7YEVL2VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TONRYHAYDM>
> .
> You are receiving this because you authored the thread.Message ID:
> <cortexrd/Knack-Toolkit-Library/repo-discussions/45/comments/7768806@
> github.com>
>


-- 
Daniel J. Kukuk
Data & Communications: AFT Michigan, AFL-CIO
***@***.***
(248) 894-9899

**Reply by CSWinnall** [reply]
You would do
_hc=[colHeader1],[ktlCond,has, colHeader1, multiSelectField,  viewIdOfMulti]
_hc=[colHeader2],[ktlCond,has, colHeader2, multiSelectField,  viewIdOfMulti]

Etc.......

**Reply by CSWinnall** [reply]
This may need to go on the details view with the multi-select field but give both places a try.  I haven't tested it.

**Reply by kukuk52** [reply]
I'm running into trouble with this. It's sometimes working with a single column but is breaking down once I add a second. I'm probably missing something. 

To test, I have a field in the grid view called 'Email' and I've got a details view on the same page displaying a single field called 'Viewable Fields' that is connected to the logged-in individual.

In the details of the grid view, I've added: _hc=[Email], [ktlCond,has, Email, Viewable Fields, view_1767]

This will work sometimes but if I continue and add something like _hc=[Department], [ktlCond,has, Department, Viewable Fields, view_1767] it breaks down. Now, both 'Email' and 'Department' are hidden regardless of if they're selected or not.  

Am I approaching this correctly? Does anything jump out as a potential solution?

**Reply by CSWinnall** [reply]
I'll set up a test to have a go myself and see if I can figure out what is wrong.

  Try using the field id instead of viewable fields just to rule out something else that may be going on. 

I'll let you know when I've tried it

Craig

**Reply by CSWinnall** [reply]
Sorry In a bit of a rush but I tried this and it worked fine :

_hc=[Client], [ktlCond,has, Client, field_540, view_212]
_hc=[Sex], [ktlCond,has, Sex, field_540, view_212]

Craig

**Reply by CSWinnall** [reply]
I can't see anything wrong or different with your _hc keywords apart from using the field id if that's the case there may be an issue with using field labels in ktlCond @cortexrd.

Let us know if that fixes it.

**Reply by kukuk52** [reply]
Yes, I was thinking of this backward (showing vs hiding). This seems to be working now! Is there a way to run the inverse of this (ie: does not contain) so that the displayed fields would be those that are selected? This just seems to be more intuitive from the user side. I'm sure we can run with this solution if that's not a possibility.

**Reply by CSWinnall** [reply]
Unfortunately I do not think that this is possible if you have all the options in one field. There is not an option for **has not** or **doesn't contain**.  

You could possibly do it if the _hc ktlCond pointed to a different field that you update with record rules depending on which options are selected not selected.
e.g. if Someone selects Email in **Viewable Fields** to show you then remove it from the other field **Hidden Fields** and vice versa.

Craig

**Reply by CSWinnall** [reply]
Yes this is doable:

Here are example keywords:

```
_hc=[Client], [ktlCond,has, Client, field_541, view_212]
_hc=[Sex], [ktlCond,has, Sex, field_541, view_212]
_hc=[Contact], [ktlCond,has, Contact, field_541, view_212]
```
Here are example record rules in your edit form:
![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/6627e8b9-f595-4b24-b454-b191d87356be)

Add 2 fields into Account
![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/5ef9675d-5273-4557-80fb-3e4b7aa6b231)

If you have a lot of fields, then this could get long but it works.

Craig

**Comment by CSWinnall** [comment]
@cortexrd is there a way to not- conditions in the ktlCond?

Like !has

**Reply by cortexrd** [reply]
Yes, I have implement this a minute ago.  
It will be **hasnt**.
I need to test it, and it should be in next release.

Norm
